### PR TITLE
z80asm: IF* pseudo ops, DEF[BW],IFEQ & instructions argument parsing, clean ups

### PR DESCRIPTION
**Pull request by sneakywumpus**
1. Added a helper function for parsing, so that DEFB, DEFW, IFEQ, IFNEQ and machine instructions now work with expressions containing ',' char constants
2. Split IF* and ELSE,ENDIF processing to make the IF* processing cleaner
3. Added rest of IF* pseudo-ops.
4. Converted opcode/operand type to flags. Added OP_NOPRE for pseudo-ops who process the operand on their own.
5. Removed TYPE, the implementation was totally wrong and it doesn't make much sense in an absolute mode only assembler.
6. NUL now works... If NUL is defined as a symbol, it won't work anymore as a unary operator, it's just too special.
6. Moved LABSEP check and label overflow processing to get_label(). I can't understand why it was in get_opcode().
7. Treat '\n' as white space, except inside strings.
8. Added NULL check to eval(), since this can happen in z80arfun.c.
9. Clean ups: '\0' and NULL comparisions, ! only for flags, "else" formatting, parentheses in "if" and "while": when necessary, not simple comparisons but more complex expressions.

The following file didn't work at all with the code before this commit:

```
XX	EQU	','
	DEFB	'A'+020H,'bcd',1,2,,4,'cd',,
	DEFW	'bc',,23,',,',,2
	IFEQ	',',XX
	XOR	A
	ELSE
	ADD	A,B
	ENDIF
	IF	NUL    '232
CCC:	LD	A,','
	ELSE
	LD	(04H+','),A
	ENDIF
	IF	NUL		
BBB:	ADD	A,B
	ENDIF
```

```
Error in file: a.asm  Line: 4
missing string delimiter
Error in file: a.asm  Line: 4
value out of range
Error in file: a.asm  Line: 9
value out of range
Error in file: a.asm  Line: 12
invalid operand
```

now

```
LOC   OBJECT CODE   LINE   STMT SOURCE CODE
002c                   1      1 XX	EQU	','
0000  61 62 63 64      2      2 	DEFB	'A'+020H,'bcd',1,2,,4,'cd',,
0004  01 02 04 63      2      3
0008  64               2      4
0009  63 62 17 00      3      5 	DEFW	'bc',,23,',,',,2
000d  2c 2c 02 00      3      6
                       4      7 	IFEQ	',',XX
0011  af               5      8 	XOR	A
                       6      9 	ELSE
                       7     10 	ADD	A,B
                       8     11 	ENDIF
                       9     12 	IF	NUL    '232
                      10     13 CCC:	LD	A,','
                      11     14 	ELSE
0012  32 30 00        12     15 	LD	(04H+','),A
                      13     16 	ENDIF
                      14     17 	IF	NUL		
0015  80              15     18 BBB:	ADD	A,B
                      16     19 	ENDIF
```

### CODE REVIEW

**Comment by ratboy666** [comment]
Thanks for your work on this assembler!

I am wondering... what assembler are you trying to follow? 

This is (almost) M80 compatible (almost) Intel compatible, and
(almost)Zilog compatible, and I have always used M80
as my reference assembler. But... just misses.

Here is the assembly attempt under M80


.Z80
0000' ASEG
ORG 0
002C XX EQU ','
0000 61 62 63 64 DEFB 'A'+020H,'bcd',1,2,,4,'cd',,
Q 0004 01 02 00 04 
Q 0008 63 64 00 00 
000C 6263 0000 DEFW 'bc',,23,',,',,2
0010 0017 2C2C 
0014 0000 0002 
U 0018 2C 2C IFEQ ',',XX
001A AF XOR A
C ELSE
LD (04H+','),A
C ENDIF
IF NUL
BBB: ADD A,B
ENDIF
END

Note the 'Q' warnings -- these are "questionable syntax".
 Specifically, warning about the ,, construct within DEFB and DEFW

Notice the trailing ,, in the DEFB... generates two (2) bytes of
zero... one for,, and one for the trailing ,.
Also, empty productions in DEFB and DEFW should evaluate to zero....
not nothing. IFEQ /IFNEQ isn't a thing with M80,
IF X EQ Y is. And really, not a thing with Zilog assemblers either --
that would be COND/ENDC

What is interesting is that, even with the errors... the results are so
close... basically, empty productions zero, and trailing
comma.

FredW

**Comment by sneakywumpus** [comment]
I mostly follow the SLR Systems Z80 macro assembler manual from 1984, but also look at the
M80 manual from 1981, and the Intel 8080/8085 Assembly Language Programming Manual from 1981,

IFEQ and IFNEQ were already part of this assembler, before I started working on it. I've added IF, COND, etc.

The SLR manual says about DEFB: "Note that a null item will generate no output, nor is it a syntax
error. This is to accommodate missing parameters in macro expansions". That sounded sensible to me.
Can't find any mentioning of empty parameters in the M80 or Intel manual regarding DB & co.

**Comment by ratboy666** [comment]
Udo, Sneakywumpus

Here is the canonical (M80, Intel) assembler expression of what you
were trying to do:

.Z80
ASEG
ORG 0
MDB MACRO A
IRP X,<A> ; NOTE THAT EACH LEVEL OF PARAMETER STRIPS OFF <> QUOTE LEVEL
IF NOT NUL X
DEFB X
ENDIF
ENDM
ENDM
MDW MACRO A
IRP X,<A>
IF NOT NUL X
DEFW X
ENDIF
ENDM
ENDM
XX EQU ','
MDB <'A'+020H,'bcd',1,2,,4,'cd',,>
MDW <'bc',,23,',,',,2>
IF ',' EQ XX
XOR A
ELSE
LD (04H+','),A
ENDIF
IF NUL
BBB: ADD A,B
ENDIF
END

Notice the use of < > quotes around arguments -- one level of this gets
stripped on each macro level expansion.
Notice the use of IF NOT NUL X to determine empty parameters and IRP
(indefinite repeat) to expand across
the list. Of course, MANY assemblers did not get this type of expansion
right... at all. The Intel assembler did,
and the Microsoft (M80) assembler did. Notice that macro expansion does
NOT occur in arguments -- 
Macros work in the text domain, purely. They do have side effects -
that can be controlled via %, which forces
the evaluation of its argument (rather than passing pure text). Couple
that with SET and LOCAL, and you
can do anything in the macro subsystem.

The Intel assembler, M80 and Digital Research MAC all had this stuff
right (not 100% sure about MAC/RMAC,
because I didn't use them "back in the day" -- it was all Intel Macro
Assembler and M80 for me).

Note the Q error is interesting. In .8080 mode:

MVI A,(LXI H)
MVI A,(SHLD)

are both very valid. This doesn't work in .Z80 mode, though. So,
sometimes there can be odd questionable
productions, especially in 8080 mode. Fewer in Z80 mode, because, even
though the assembler
APPEARS more othogonal, it is actually less consistent. Just some "food
for thought".

FredW

**Comment by udo-munk** [comment]
Originally the assembler was a substitute for Mostek's Z80 cross assembler for UNIX, so that is were IFEQ and IFNEQ came from, because I needed to assemble sources written for that without changes. This is not relevant anymore nowadays.

**Comment by ratboy666** [comment]
And, as a PS.

TYPE makes sense, even in an absolute only assembler. 20h is set if the
expression
can be evaluated and is NOT external. If it can be evaluated and is
external,
it would be 80h set. If neither of these two bits are set, it isn't a
valid expression.

This can be very valuable information in a macro... It can be used like
this:

macro a
  if type x and 20h
    db x
  else
    x
 endif
endm

y macro
  db 'hey hey'
endm

 a 2
  a y

See? we use type here to distinguish between an evaluable expression
and a futher
macro call. And that is why it is valuable, even in a non-relocatable
scenario.

FredW

Basically, if type x and 20h is true, then %x will also work! No error,
no warning, no foul.

**Comment by ratboy666** [comment]
Got IFEQ now! That makes sense. I would do this:

IFEQ MACRO X,Y
  IF X EQ Y
ENDM

(and of course, the other for IFNEQ) -- that should replace IFEQ and IFNEQ nicely,
without otherwise having them in the assembler itself.

Works with M80,

**Comment by ratboy666** [comment]
And, notice the blank item detection in my MDB and MDW macros. These can replace DEFB and DEFW without error in M80. Just use DB and DW in the macro expansion itself (otherwise you
get infinite recursion).

So:

DEFB MACRO A
  IRP X,<A> ; NOTE THAT EACH LEVEL OF PARAMETER STRIPS OFF <> QUOTE LEVEL
  IF NOT NUL X
  DB X
  ENDIF
  ENDM
ENDM

and

DEFW MACRO A
  IRP X,<A>
  IF NOT NUL X
  DW X
  ENDIF
  ENDM
ENDM

The only care is that the argument lists to DEFB and DEFW need to be in < > brackets.

This also works with M80 (not able to test old Intel assembler anymore).

**Comment by ratboy666** [comment]
And, as I remember, the Mostek assembler kind of sucked ass when trying to do advanced
Macro work. We would have used a separate macro expansion phase for that, or just skipped
that assembler. sorry that the editor here stripped out less than/greater than signs...

**Comment by ratboy666** [comment]
% is described on 4-8 of the M80 manual. NUL (and the Microsoft preferred IFB and IFNB)
are on page 4-7. Note that the Intel assembler did not have IFB and IFNB, just NUL. These
things "complete" the macro capabilities of the assembler, making it possible to do high level
code generations: This can get very close to BASIC or COBOL

**Comment by udo-munk** [comment]
We were doing very advanced Macro work with the Mostek assembler, but it was tricky. Also we made changes to the assembler, we got the source code from Mostek. Whatever, not relevant anymore and not necessary to keep compatibility to it.
